Add information-only thread naming

kcgen opened the pull request:
The autotype, and MT-32 and FluidSynth rendering threads are named.

For example:

![2021-01-30_13-23](https://user-images.githubusercontent.com/1557255/106368510-d504aa00-62fe-11eb-8e08-ab29f424b620.png)

To test this, start autotype for a long enough sequence that your process monitor (htop, top, task manager) can pick it up.

Note that the autotype thread will disappear from the list once autotyping is complete.

dreamer commented:
Without doing the proper review yet… We wouldn't need to introduce these functions if we stuck to using SDL2 threads… They provide several functionalities we like, and thread naming is one of them. So perhaps it would be better to migrate std::thread to SDL_thread?

kcgen commented:
C++11 threads are literally part of the language spec, so we get scoped locks and can more easily  without the explicit C-based destroy calls (like SDL thread destroy, SDL condition destroy, SDL mutex destroy). 

So for general threading needs, I see all of the above providing advantages, and I see the extra C-based SDL baggage as disadvantages. 

I also prefer we stick to vanilla language calls when we don't have a specific requirement to use a frameworks' constructs.  

For example, if we now feel we should prefer SDL constructs in a general sense over C++11, then we might as well start using `SInt32` and `SDL_Bool` instead of `int32_t` and `bool`.

SDL does offer thread naming, but I see this as a single-call nice-to-have, and of less value than the points above. Given the two options, I will always prefer `std::thread`s. 

That does bring up and interesting point: given SDL is simply making these same OS-specific naming calls under the hood, maybe there's a way I can just call their code inside `set_thread_name`.

kcgen commented:
**SDL_SYS_SetupThread** is how SDL names their threads:

``` text
src/thread/SDL_systhread.h:extern void SDL_SYS_SetupThread(const char *name);
src/thread/SDL_thread.c:    SDL_SYS_SetupThread(thread->name);
src/thread/generic/SDL_systhread.c:SDL_SYS_SetupThread(const char *name)
src/thread/os2/SDL_systhread.c:SDL_SYS_SetupThread(const char *name)
src/thread/psp/SDL_systhread.c:void SDL_SYS_SetupThread(const char *name)
src/thread/pthread/SDL_systhread.c:SDL_SYS_SetupThread(const char *name)
src/thread/stdcpp/SDL_systhread.cpp:SDL_SYS_SetupThread(const char *name)
src/thread/windows/SDL_systhread.c:SDL_SYS_SetupThread(const char *name)
```

And the symbol is available in the static library `libSDL2.a` but not in the shared library - so I can't leverage it:
```
                 U SDL_SYS_SetupThread
0000000000000110 T SDL_SYS_SetupThread
```

Thinking about the code in this PR, the Windows chuck is quite ugly - so perhaps we could just drop that all together and only provide naming on non-macOS posix systems, where it's going to be a bit more interesting to those users.

dreamer commented:
- Yes, dropping Windows-only portion of this implementation (which uses deprecated Windows calls BTW) would be preferred to merging it as-is.
- `SDL_SYS_SetupThread` is most likely a private SDL abstraction around various thread implementations, no wonder it's hidden from end users
- mapping to thread ids would indeed not show up in top/htop (but most likely such solution is good enough to for using in logs, to identify which thread is doing what)

There's one more option, that we haven't discussed yet - we can develop our own wrapper mimicking `std::thread` providing C++ abstraction over SDL_thread (e.g. class named `sdl::thread` having very similar interface to `std::thread`, but abstracting SDL_thread instead of POSIX threads/Windows/other threads). In one of other projects I worked on my team developed such shallow abstraction (over yet another threading API), and it worked very well. Let me dig a bit, maybe I'll find the code - AFAIR implementing such abstraction required learning few quirks of STL interface, but overall was rather easy.

dreamer commented:
Some time ago I started writing a small wrapper for SDL_thread, but haven't finished it: https://gist.github.com/dreamer/9898fd4d5ac869e358ff75891712ab86 There are other such wrappers available, mostly from C++0x times - I looked at their implementation and did not like them very much. I don't feel like continuing work on this wrapper, and I probably won't maintain it. It's usable, but in an unfinished state (not even hard to finish - I just don't feel like writing C++ wrappers if I don't have to).

Back to topic of this PR. We have two motivations for preferring `SDL_thread` over `std::thread`:

- ability to reconfigure thread priority
  - for mt32emu (but it seems like with new multithreaded implementation, changing priority is not needed any more)
  - for fluidsynth (same as mt32emu)
  - for `priority` setting in `[sdl]` section (but I think we'll be more likely to remove this setting altogether)
- thread naming - I don't think it's necessary, and implementation in this PR is not good enough for general use (SDL implements hidden thread naming storage to work around paltform-specific limitations). It is good enough for "informational" use, therefore I suggest to trim it further and *remove Windows implementation completely*. POSIX thread naming works on macOS (despite what implementation says), so this would need to be adjusted as well.

@kcgen Is it ok if we'll limit this implementation to being "informational only"? I mean: it will be possible to set the thread name in code, but impossible to distinguish threads by their names.
- If answer is yes, then I'll continue with review with this usage in mind. 
- If the answer is no, then moving forward with `sdl_thread` wrapper will be the only acceptable option.

kcgen commented:
> Is it ok if we'll limit this implementation to being "informational only"?

Absolutely!

>  trim it further and remove Windows implementation completely. POSIX thread naming works on macOS (despite what implementation says), so this would need to be adjusted as well.

Will do.

kcgen commented:
Updated and added to Meson; curious what you think. 
(Checking for the function required `threads_dep` :sweat_smile:)

Here are the two queues in action.

### MT-32

![2021-03-11_13-53](https://user-images.githubusercontent.com/1557255/110860234-5ee65200-8271-11eb-8b37-94dcb1e27fc8.png)

### FluidSynth

![2021-03-11_13-54](https://user-images.githubusercontent.com/1557255/110860236-5f7ee880-8271-11eb-9712-4409a15b2967.png)

dreamer commented on src/misc/support.cpp:
```diff
@@ -331,3 +331,15 @@ std::string safe_strerror(int err) noexcept
+	pthread_t handle = thread.native_handle();
+	pthread_setname_np(handle, name);
+#else
+	// Do-nothing if platform can't name threads
+	(void)thread;
+	(void)name;
```
Remove these lines and use mark the parameters with `MAYBE_UNUSED` instead. This way it will be easier for me to mechanically replace `MAYBE_UNUSED` with C++17 `[[maybe_unused]]` in the future.

Clang is silly in warning about using C++17 extensions (we specifically ask compiler if attributes are supported before using them).

kcgen replied:
Got it. Yeah; I didn't know which approach to use given in some cases the variables are 100% used and in others they're not.  Lol.. I suppose that's the meaning of `MAYBE_UNUSED` :smile:

dreamer commented on src/misc/support.cpp:
```diff
@@ -331,3 +331,15 @@ std::string safe_strerror(int err) noexcept
+{
+#if HAVE_PTHREAD_SETNAME_NP
+	pthread_t handle = thread.native_handle();
+	pthread_setname_np(handle, name);
```
You should add an assert before calling `pthread_setname_np`:
```C++
assert(strlen(name) < 16);
```

kcgen replied:
Great suggestion; thanks @dreamer.

dreamer commented on src/gui/sdl_mapper.cpp:
```diff
@@ -1228,7 +1228,8 @@ class Typer {
 			m_stop_requested = false;
 			m_instance = std::thread(&Typer::Callback, this);
-		}
+		        set_thread_name(m_instance, "dosbox-autotype");
```
Whitespace.

kcgen replied:
Interesting - this is somehow confusing clang-format.
Even when I manually correct it with tabs, clang-format switches it back to `<tab><tab> + 8-spaces`.
(manually fixing - thanks for catching it!)

dreamer reviewed (changes requested):
Where does `dosbox:disk$1` process come from?

kcgen commented on meson.build:
```diff
@@ -252,6 +252,21 @@ if host_machine.system() in ['windows', 'cygwin']
   winmm_dep = cxx.find_library('winmm', required : true)
 endif
 
+# For GNU extensions, we need to define _GNU_SOURCE and tell Meson to
+# compile and link with thread_dep dependency when checking this feature.
+if cc.links('''
+            #ifndef _GNU_SOURCE
+            #define _GNU_SOURCE
+            #endif
+            #include <pthread.h>
+            int main() {
+                return pthread_setname_np(pthread_self(), "test");
+            }
+            ''',
+            name : 'pthread_setname_np',
+            dependencies : threads_dep)
+    conf_data.set10('HAVE_PTHREAD_SETNAME_NP', true)
+endif
```
> `dependencies : dependency('threads'))`

:arrow_up: That's what I was missing when I originally tried:
   `cc.has_function('pthread_setname_np', '#include <pthread.h>')`

Can we be more specific about the function we test for?

On macOS, `pthread.h` only supports setting the name when inside the thread:

``` c++
int pthread_setname_np(const char*);
```

I'm worried this check will return true in this case, but it's not the function we want.

kcgen replied:
I've been combing the meson documentation but can't see any other way (besides running a link test) to check for the presence of a function _and_ its arguments_ and types.

So without a link-check, we get this on macOS:

![2021-03-12_09-27](https://user-images.githubusercontent.com/1557255/110976025-40399700-8315-11eb-98ff-03f4e3e76701.png)

Should I exclude `MACOSX` on the source-side?  

``` c++
#if HAVE_PTHREAD_SETNAME_NP && !defined(MACOSX)
```

kcgen replied:
**Edit** instead of excluding MacOS, I opted to turn this into a positive check for GNU extensions, which we also use in other support functions.

``` c++
#if defined(HAVE_PTHREAD_SETNAME_NP) && defined(_GNU_SOURCE)
```

kcgen commented:
> Where does `dosbox:disk$0`..`dosbox:disk$3` process names come from?

They come from MESA .. 

``` c++
   /* 4 threads were chosen below because just about all modern CPUs currently
    * available that run Mesa have *at least* 4 cores. For these CPUs allowing
    * more threads can result in the queue being processed faster, thus
    * avoiding excessive memory use due to a backlog of cache entrys building
    * up in the queue. Since we set the UTIL_QUEUE_INIT_USE_MINIMUM_PRIORITY
    * flag this should have little negative impact on low core systems.
    *
    * The queue will resize automatically when it's full, so adding new jobs
    * doesn't stall.
    */
   if (!util_queue_init(&cache->cache_queue, "disk$", 32, 4,
                        UTIL_QUEUE_INIT_RESIZE_IF_FULL |
                        UTIL_QUEUE_INIT_USE_MINIMUM_PRIORITY |
                        UTIL_QUEUE_INIT_SET_FULL_THREAD_AFFINITY))
```

https://sources.debian.org/src/mesa/21.0.0%7Erc5-1/src/util/disk_cache.c/#L155

kcgen commented:
> Where does `dosbox:disk$0`..`dosbox:disk$3` process names come from?

They come from MESA .. 

``` c++
   /* 4 threads were chosen below because just about all modern CPUs currently
    * available that run Mesa have *at least* 4 cores. For these CPUs allowing
    * more threads can result in the queue being processed faster, thus
    * avoiding excessive memory use due to a backlog of cache entrys building
    * up in the queue. Since we set the UTIL_QUEUE_INIT_USE_MINIMUM_PRIORITY
    * flag this should have little negative impact on low core systems.
    *
    * The queue will resize automatically when it's full, so adding new jobs
    * doesn't stall.
    */
   if (!util_queue_init(&cache->cache_queue, "disk$", 32, 4,
                        UTIL_QUEUE_INIT_RESIZE_IF_FULL |
                        UTIL_QUEUE_INIT_USE_MINIMUM_PRIORITY |
                        UTIL_QUEUE_INIT_SET_FULL_THREAD_AFFINITY))
```

https://sources.debian.org/src/mesa/21.0.0%7Erc5-1/src/util/disk_cache.c/#L155

dreamer reviewed (approved):
D'oh I haven't noticed that macOS version of `pthread_setname_np` is different than GNU `pthread_setname_np`… I guess that explains why SDL allows changing thread name on creation only…

Nice work - merging it in, thanks! :)
